File: examples/anchor-multi-program/tests/test_cpi.ts
import { AnchorProvider, Program, Wallet, web3, BN } from "@coral-xyz/anchor";
import { LiteSVM } from "../local-litesvm";

import A_IDL from "../target/idl/program_a.json";
import B_IDL from "../target/idl/program_b.json";
import {
  Keypair,
  LAMPORTS_PER_SOL,
  PublicKey,
} from "@solana/web3.js";
import { ProgramA} from "../target/types/program_a";
import { ProgramB } from "../target/types/program_b";
import { assert } from "console";

describe("anchor multi program, litesvm tests", () => {
    let svm: LiteSVM;
    let payer: Keypair;
    let program_a: Program<ProgramA>;
    let program_a_Id: PublicKey;
    let program_b: Program<ProgramB>;
    let program_b_Id: PublicKey;

    before(async () => {
        svm = new LiteSVM();
        payer = new Keypair();
        svm.airdrop(payer.publicKey, BigInt(LAMPORTS_PER_SOL));

        const provider = new AnchorProvider(
            svm as any, 
            new Wallet(payer), 
            {
                commitment: "confirmed",
            }
        );

        program_a = new Program(A_IDL, provider);
        // program_a_Id = new web3.PublicKey('2BJiU3UUhRmroYHXN6iEcbuw7PfDAJqcFRv9AFutQxzQ');
        program_a_Id = program_a.programId;

        svm.addProgramFromFile(program_a_Id, "target/deploy/program_a.so");

        program_b = new Program(B_IDL, provider);
        // program_b_Id = new web3.PublicKey('G8pztLe793X3k3u8FGVdyM9aBaqY8UPYuknzKVyay9N4');
        program_b_Id = program_b.programId;

        svm.addProgramFromFile(program_b_Id, "target/deploy/program_b.so");
    })

    it("test_cpi_from_program_a_to_program_b", async () => {
        console.log("Program A ID:", program_a_Id.toBase58());
        console.log("Program B ID:", program_b_Id.toBase58());
        const cpi_instruction = await program_a.methods
            .cpi()
            .accounts({
                programB: program_b_Id,
            })
            .instruction();

        const blockhash = svm.latestBlockhash();
        const tx = new web3.Transaction();

        tx.recentBlockhash = blockhash;
        tx.add(cpi_instruction);
        tx.sign(payer);
    
        const result = svm.sendTransaction(tx);
        console.log("Transaction result:", JSON.stringify(result, null, 2));
    });
})

File: examples/anchor-multi-program/programs/program-a/tests/test_cpi.rs
use litesvm::LiteSVM;
use solana_instruction::{AccountMeta, Instruction};
use solana_keypair::Keypair;
use solana_message::Message;
use solana_pubkey::pubkey;
use solana_sdk_ids::system_program;
use solana_signer::Signer;
use solana_transaction::Transaction;

#[test]
fn test_cpi() {
    let mut svm: LiteSVM = LiteSVM::new();


    let program_id = pubkey!("ALkpavDZFcbRNjZn8cf3ptLfT9DSYQ3hkKBxeZPAXYz2");
    svm.add_program_from_file(program_id, "../../target/deploy/program_a.so")
        .unwrap();

    // Load program B
    let program_b_id = pubkey!("G8pztLe793X3k3u8FGVdyM9aBaqY8UPYuknzKVyay9N4");
    svm.add_program_from_file(program_b_id, "../../target/deploy/program_b.so")
        .unwrap();

    let signer_keypair = Keypair::new();
    let signer_pubkey = signer_keypair.pubkey();
    svm.airdrop(&signer_pubkey, 10_000_000).unwrap();

    // 1. This is the IDL of the cpi method in program A - see programs/program-a/src/lib.rs -> Discriminator of the cpi method
    let ix_data = vec![76, 173, 6, 95, 181, 93, 83, 206];

    // 2. Create the Accounts
    let accounts_ix = vec![
        // AccountMeta::new(signer_pubkey, true),
        AccountMeta::new(program_b_id, false),
        // AccountMeta::new_readonly(system_program::ID, false),
    ];

    // 3. Build instruction for program_a
    let instruction_a = Instruction {
        program_id,
        accounts: accounts_ix,
        data: ix_data,
    };

    // 5. Build and send transaction with both instructions
    let message = Message::new(&[instruction_a.clone()], Some(&signer_pubkey));
    let tx = Transaction::new(&[&signer_keypair], message, svm.latest_blockhash());

    let result = svm.send_transaction(tx);
    assert!(result.is_ok(), "Transaction failed: {:#?}", result.err()); // TODO: IMPORTANT usage of `#` will format the error
}

fn test_non_cpi() {
    let mut svm: LiteSVM = LiteSVM::new();

    let program_id = pubkey!("2BJiU3UUhRmroYHXN6iEcbuw7PfDAJqcFRv9AFutQxzQ");
    svm.add_program_from_file(program_id, "../../target/deploy/program_a.so")
        .unwrap();

    let signer_keypair = Keypair::new();
    let signer_pubkey = signer_keypair.pubkey();
    svm.airdrop(&signer_pubkey, 10_000_000).unwrap();

    let accounts_ix = vec![];

    let ix_data = vec![86, 36, 10, 211, 246, 235, 42, 57];

    let instruction_a = Instruction {
        program_id,
        accounts: accounts_ix,
        data: ix_data,
    };

    let message = Message::new(&[instruction_a], Some(&signer_pubkey));
    let tx = Transaction::new(&[&signer_keypair], message, svm.latest_blockhash());

    let result = svm.send_transaction(tx);
    assert!(result.is_ok(), "Transaction failed: {:#?}", result.err()); // TODO: IMPORTANT usage of `#` will format the error
}
